Use codegen'ed unboxing wrappers

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33905 Refactor jit::Operator
* #33904 Fix either::map naming
* #33313 Improve boxed dispatch performance
* **#32521 Use codegen'ed unboxing wrappers**
* #33732 Remove unnecessary tensor copies

Not all ops support the templated unboxing wrappers yet. For the ones that don't,
let's use the codegen'ed unboxing wrappers from register_aten_ops.cpp, but register
them with c10 directly instead of JIT.

The `use_c10_dispatcher` setting in `native_functions.yaml` now has a new option 'with_codegenerated_unboxing_wrapper' which means we take the codegened unboxing wrapper from register_aten_ops.cpp and stuff it into c10. This new argument is the default, 'unboxed_only' is not the default anymore. For the (very few) ops that don't support boxed dispatch yet (i.e. ops taking TensorOptions arguments), we set them to 'unboxed_only' and they follow the old behavior of having register_aten_ops.cpp register the jit op.

Next steps here are (1) to make TensorOptions work with boxed dispatch and remove the `unboxed_only` option from `use_c10_dispatcher`, so that all ops go through the new path and (2) make the new path template-only and remove codegen from it (see https://github.com/pytorch/pytorch/issues/32366).

Benchmarks
-------
There were some regressions visible from this originally, but after stacking this on top of https://github.com/pytorch/pytorch/pull/33732, there's no regressions anymore.

|  | add | avg1dpool | resnet |
|---|---|---|---|
| before this PR (i.e. https://github.com/pytorch/pytorch/pull/33732) | 1.66us | 5.83us | 103.61us | 
| after this PR | 1.66us | 6.09us | 103.95us | 
| after this PR and https://github.com/pytorch/pytorch/pull/33313 | | 5.74us | |

When we disable variable dispatch, we get the following numbers:

|  | add | avg1dpool | resnet |
|---|---|---|---|
| before this PR (i.e. https://github.com/pytorch/pytorch/pull/33732) | 1.55us | 4.11us | 103.36us | 
| after this PR | 1.57us | 4.23us | 102.55us | 
| after this PR and https://github.com/pytorch/pytorch/pull/33313 | | 3.84us | |


Differential Revision: [D18672405](https://our.internmc.facebook.com/intern/diff/D18672405/)

smessmer committed:
[wip] Use codegen'ed unboxing wrappers

Not all ops support the templated unboxing wrappers yet. For the ones that don't,
let's use the codegen'ed unboxing wrappers from register_aten_ops.cpp, but register
them with c10 directly instead of JIT.

The `use_c10_dispatcher` setting in `native_functions.yaml` now has a new option 'with_codegenerated_unboxing_wrapper' which means we take the codegened unboxing wrapper from register_aten_ops.cpp and stuff it into c10. This new argument is the default, 'unboxed_only' is not the default anymore. For the ops that don't support boxed dispatch yet (i.e. ops taking TensorOptions arguments), we set them to 'unboxed_only' and they follow the old behavior of having register_aten_ops.cpp register the jit op.

This temporary hack will allow us to get mobile working much faster, but the plan is
to remove this hack again once all ops support the templated unboxing wrapper.

Differential Revision: [D18672405](https://our.internmc.facebook.com/intern/diff/D18672405/)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 262755d1:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=262755d17392f22e24c34971c802af11c6f48fd7).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4322335) pytorch_linux_backward_compatibility_check_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=89691673) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=89691673))
<details>
<summary>
<code>Jan 23 04:38:40 The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not.</code>
</summary>

```
Jan 23 04:38:40 processing existing schema:  aten::values(Dict(str, t) self) -> (t[](*)) 
Jan 23 04:38:40 processing existing schema:  aten::values(Dict(int, t) self) -> (t[](*)) 
Jan 23 04:38:40 processing existing schema:  aten::values(Dict(float, t) self) -> (t[](*)) 
Jan 23 04:38:40 processing existing schema:  aten::values(Dict(Tensor, t) self) -> (t[](*)) 
Jan 23 04:38:40 processing existing schema:  aten::view(Tensor(a) self, int[] size) -> (Tensor(a)) 
Jan 23 04:38:40 processing existing schema:  aten::where(Tensor condition) -> (Tensor[]) 
Jan 23 04:38:40 processing existing schema:  aten::where.self(Tensor condition, Tensor self, Tensor other) -> (Tensor) 
Jan 23 04:38:40 processing existing schema:  aten::zeros_like(Tensor self, *, int? memory_format=None) -> (Tensor) 
Jan 23 04:38:40 processing existing schema:  aten::zeros_like.dtype(Tensor self, *, int dtype, int layout, Device device, bool pin_memory=False, int? memory_format=None) -> (Tensor) 
Jan 23 04:38:40 processing existing schema:  prim::Constant() -> (...) 
Jan 23 04:38:40 The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not.  
Jan 23 04:38:40  
Jan 23 04:38:40 Broken ops: [ 
Jan 23 04:38:40 	aten::index_put(Tensor self, Tensor[] indices, Tensor values, bool accumulate=False) -> (Tensor) 
Jan 23 04:38:40 	aten::index.Tensor(Tensor self, Tensor[] indices) -> (Tensor) 
Jan 23 04:38:40 	aten::_index_put_impl_(Tensor(a!) self, Tensor[] indices, Tensor values, bool accumulate=False, bool unsafe=False) -> (Tensor(a!)) 
Jan 23 04:38:40 	aten::index_put_(Tensor(a!) self, Tensor[] indices, Tensor values, bool accumulate=False) -> (Tensor(a!)) 
Jan 23 04:38:40 ] 
Jan 23 04:38:40 + cleanup 
Jan 23 04:38:40 + retcode=1 
Jan 23 04:38:40 + set +x 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

smessmer committed:
Update on "[wip] Use codegen'ed unboxing wrappers"


Not all ops support the templated unboxing wrappers yet. For the ones that don't,
let's use the codegen'ed unboxing wrappers from register_aten_ops.cpp, but register
them with c10 directly instead of JIT.

The `use_c10_dispatcher` setting in `native_functions.yaml` now has a new option 'with_codegenerated_unboxing_wrapper' which means we take the codegened unboxing wrapper from register_aten_ops.cpp and stuff it into c10. This new argument is the default, 'unboxed_only' is not the default anymore. For the ops that don't support boxed dispatch yet (i.e. ops taking TensorOptions arguments), we set them to 'unboxed_only' and they follow the old behavior of having register_aten_ops.cpp register the jit op.

This temporary hack will allow us to get mobile working much faster, but the plan is
to remove this hack again once all ops support the templated unboxing wrapper.

Differential Revision: [D18672405](https://our.internmc.facebook.com/intern/diff/D18672405/)

[ghstack-poisoned]

smessmer committed:
Update on "Use codegen'ed unboxing wrappers"


Not all ops support the templated unboxing wrappers yet. For the ones that don't,
let's use the codegen'ed unboxing wrappers from register_aten_ops.cpp, but register
them with c10 directly instead of JIT.

The `use_c10_dispatcher` setting in `native_functions.yaml` now has a new option 'with_codegenerated_unboxing_wrapper' which means we take the codegened unboxing wrapper from register_aten_ops.cpp and stuff it into c10. This new argument is the default, 'unboxed_only' is not the default anymore. For the (very few) ops that don't support boxed dispatch yet (i.e. ops taking TensorOptions arguments), we set them to 'unboxed_only' and they follow the old behavior of having register_aten_ops.cpp register the jit op.

Next steps here are (1) to make TensorOptions work with boxed dispatch and remove the `unboxed_only` option from `use_c10_dispatcher`, so that all ops go through the new path and (2) make the new path template-only and remove codegen from it (see https://github.com/pytorch/pytorch/issues/32366).

First experiments show that
- For a small JITted model that calls add (i.e. a op with just two arguments that are both tensors) on two tensors in a loop, we see a 2-4% performance improvement (~35-50ns) when compared to the old path. This is a simple op that takes two tensor arguments and no non-tensor arguments, so iterating over it in boxed dispatch is cheap.
- For a small JITted model that calls avgpool1d (i.e. an op that has one tensor arg and 5 non-tensor args) on a tensor in a loop, we see a 3-4% performance regression (~60ns) when compared to the old path. This is an op that takes only one tensor argument and then 6 non-tensor arguments. Unboxed dispatch doesn’t have to look at those but boxed dispatch still needs to iterate over them.

This performance difference is likely due to boxed dispatch iterating over all arguments in a loop and unboxed dispatch not having to look at non-tensor arguments.

Differential Revision: [D18672405](https://our.internmc.facebook.com/intern/diff/D18672405/)

[ghstack-poisoned]

zdevito commented on tools/jit/gen_jit_dispatch.py:
```diff
@@ -132,6 +132,13 @@ def from_ivalue(arg, value):
 
 
+CALL_UNBOXED_KERNEL = CodeTemplate("""\
+using FuncType = ${return_type} (${formals_types});
```
Can this logic be put in a single-line function and not generated? It looks like that founction would be pretty simple

smessmer replied:
This is supposed to replace CALL_NAMESPACE, CALL_METHOD, CALL_NAMESPACE_WITH_TENSOR_OPTIONS and CALL_METHOD_WITH_TENSOR_OPTIONS. That's why, instead of having those all call a common function, I chose to keep this separate.

bhosmer replied:
Sorry to be dense but I'm not sure I get the connection... why would replacing* those things make a common function undesirable?

*in the future

smessmer replied:
ah I misunderstood the comment. Made it into a function now.

zdevito commented on tools/jit/gen_jit_dispatch.py:
```diff
@@ -302,7 +321,24 @@ def pack_arguments(args):
+            elif len(decl['returns']) == 1:
+                return_type = decl['returns'][0]['type']
+            else:
+                return_type = "std::tuple<" + ", ".join([r['type'] for r in decl['returns']]) + ">"
```
Use .format plase, this is hard to read.

zdevito reviewed (commented):
This is conditionally good with me but I want @bhosmer to look at the code generation changes as well. This PR temporarily introduces even more complexity into the operator library, so it is important that this complexity goes away eventually

Conditions
* Landing this is a commitment to removing the setManuallyBoxedKernel hacks. When landing please give an estimated date when these should be gone so we can schedule a followup.
* This should be benchmarked to check that it doesn't regress performance of benchmark models noticeable, and does not cause significant (e.g. > 30%) overhead to dispatch of these operators individually.

dzhulgakov reviewed (commented):
Great progress!

I think when we get rid of `unboxed_only` the code would actually simplify quite a bit. Agree with Zach - if we're confident that it happens soon and perf is not an issue, I'm fine with landing this PR

bhosmer reviewed (commented):
Some minor comments inline, will continue reviewing today

smessmer committed:
Update on "Use codegen'ed unboxing wrappers"


Not all ops support the templated unboxing wrappers yet. For the ones that don't,
let's use the codegen'ed unboxing wrappers from register_aten_ops.cpp, but register
them with c10 directly instead of JIT.

The `use_c10_dispatcher` setting in `native_functions.yaml` now has a new option 'with_codegenerated_unboxing_wrapper' which means we take the codegened unboxing wrapper from register_aten_ops.cpp and stuff it into c10. This new argument is the default, 'unboxed_only' is not the default anymore. For the (very few) ops that don't support boxed dispatch yet (i.e. ops taking TensorOptions arguments), we set them to 'unboxed_only' and they follow the old behavior of having register_aten_ops.cpp register the jit op.

Next steps here are (1) to make TensorOptions work with boxed dispatch and remove the `unboxed_only` option from `use_c10_dispatcher`, so that all ops go through the new path and (2) make the new path template-only and remove codegen from it (see https://github.com/pytorch/pytorch/issues/32366).

Benchmarks
-------

|  | add | avg1dpool | resnet |
|---|---|---|---|
| before this PR | 1.75us | 6.06us | 103.62us | 
| after this PR | 1.86us | 6.05us | 103.72us | 
| after this PR + performance fixes | 1.88us | 5.68us | 103.98us | 

When we disable variable dispatch, we get the following numbers:

|  | add | avg1dpool | resnet |
|---|---|---|---|
| before this PR | 1.59us | 4.01us | 104.20us | 
| after this PR | 1.71us | 4.20us | 103.30us | 
| after this PR + performance fixes | 1.67us | 3.91us | 102.99us | 

We see a slight performance decrease for the `add()` operator (which is an operator taking two tensor arguments) but no meaningful difference on avgpool1d (which is an operator taking a lot of non-tensor arguments) or on an actual model.


Differential Revision: [D18672405](https://our.internmc.facebook.com/intern/diff/D18672405/)

[ghstack-poisoned]

bhosmer commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -52,17 +53,20 @@
 - func: set_data(Tensor(a!) self, Tensor new_data) -> ()
   use_c10_dispatcher: full
   manual_kernel_registration: True
+  use_c10_dispatcher: unboxed_only
```
Several of these have double entries for use_c10_dispatcher FYI

smessmer committed:
Update on "Use codegen'ed unboxing wrappers"


Not all ops support the templated unboxing wrappers yet. For the ones that don't,
let's use the codegen'ed unboxing wrappers from register_aten_ops.cpp, but register
them with c10 directly instead of JIT.

The `use_c10_dispatcher` setting in `native_functions.yaml` now has a new option 'with_codegenerated_unboxing_wrapper' which means we take the codegened unboxing wrapper from register_aten_ops.cpp and stuff it into c10. This new argument is the default, 'unboxed_only' is not the default anymore. For the (very few) ops that don't support boxed dispatch yet (i.e. ops taking TensorOptions arguments), we set them to 'unboxed_only' and they follow the old behavior of having register_aten_ops.cpp register the jit op.

Next steps here are (1) to make TensorOptions work with boxed dispatch and remove the `unboxed_only` option from `use_c10_dispatcher`, so that all ops go through the new path and (2) make the new path template-only and remove codegen from it (see https://github.com/pytorch/pytorch/issues/32366).

Benchmarks
-------

|  | add | avg1dpool | resnet |
|---|---|---|---|
| before this PR | 1.75us | 6.06us | 103.62us | 
| after this PR | 1.86us | 6.05us | 103.72us | 
| after this PR + performance fixes | 1.88us | 5.68us | 103.98us | 

When we disable variable dispatch, we get the following numbers:

|  | add | avg1dpool | resnet |
|---|---|---|---|
| before this PR | 1.59us | 4.01us | 104.20us | 
| after this PR | 1.71us | 4.20us | 103.30us | 
| after this PR + performance fixes | 1.67us | 3.91us | 102.99us | 

We see a slight performance decrease for the `add()` operator (which is an operator taking two tensor arguments) but no meaningful difference on avgpool1d (which is an operator taking a lot of non-tensor arguments) or on an actual model.


Differential Revision: [D18672405](https://our.internmc.facebook.com/intern/diff/D18672405/)

[ghstack-poisoned]

smessmer commented on tools/jit/gen_jit_dispatch.py:
```diff
@@ -159,26 +166,38 @@ def from_ivalue(arg, value):
 
 CONSTRUCTOR = CodeTemplate("""\
-[](Stack & stack) {
+[](OperatorKernel* unboxedKernel, const OperatorHandle&, Stack* stack) {
```
c10 and jit represent those differently, we should unify at some point in the future

smessmer commented:
`setManuallyBoxedKernel` can be deleted once https://github.com/pytorch/pytorch/issues/32366 is fixed

smessmer committed:
Update on "Use codegen'ed unboxing wrappers"


Not all ops support the templated unboxing wrappers yet. For the ones that don't,
let's use the codegen'ed unboxing wrappers from register_aten_ops.cpp, but register
them with c10 directly instead of JIT.

The `use_c10_dispatcher` setting in `native_functions.yaml` now has a new option 'with_codegenerated_unboxing_wrapper' which means we take the codegened unboxing wrapper from register_aten_ops.cpp and stuff it into c10. This new argument is the default, 'unboxed_only' is not the default anymore. For the (very few) ops that don't support boxed dispatch yet (i.e. ops taking TensorOptions arguments), we set them to 'unboxed_only' and they follow the old behavior of having register_aten_ops.cpp register the jit op.

Next steps here are (1) to make TensorOptions work with boxed dispatch and remove the `unboxed_only` option from `use_c10_dispatcher`, so that all ops go through the new path and (2) make the new path template-only and remove codegen from it (see https://github.com/pytorch/pytorch/issues/32366).

Benchmarks
-------
There were some regressions visible from this originally, but after stacking this on top of https://github.com/pytorch/pytorch/pull/33732, there's no regressions anymore.

|  | add | avg1dpool | resnet |
|---|---|---|---|
| before this PR (i.e. https://github.com/pytorch/pytorch/pull/33732) | 1.66us | 5.83us | 103.61us | 
| after this PR | 1.66us | 6.09us | 103.95us | 
| after this PR and https://github.com/pytorch/pytorch/pull/33313 | | 5.74us | |

When we disable variable dispatch, we get the following numbers:

|  | add | avg1dpool | resnet |
|---|---|---|---|
| before this PR (i.e. https://github.com/pytorch/pytorch/pull/33732) | 1.55us | 4.11us | 103.36us | 
| after this PR | 1.57us | 4.23us | 102.55us | 
| after this PR and https://github.com/pytorch/pytorch/pull/33313 | | 3.84us | |


Differential Revision: [D18672405](https://our.internmc.facebook.com/intern/diff/D18672405/)

[ghstack-poisoned]

bhosmer commented on tools/jit/gen_jit_dispatch.py:
```diff
@@ -518,6 +572,12 @@ def needs_hacked_twin(decl):
 def hacked_twin(decl):
     decl_copy = copy.deepcopy(decl)
     decl_copy['schema_string'] = decl['schema_string'].replace('Tensor?[]', 'Tensor[]')
+    if decl['overload_name']:
```
nit: these lines are kind of long and busy and we're dereferencing stuff out of `decl`/`decl_copy` over and over... if you get the chance before landing, hoisting out some locals should help one and eliminate the other

bhosmer replied:
Also, as background, are you fixing a bug here with the overload name dedup? Or is there something about the codegen move that necessitates it?

smessmer replied:
For background: This was a bug but it didn't materialize before

bhosmer commented on torch/csrc/jit/register_c10_ops.cpp:
```diff
@@ -13,7 +13,7 @@ namespace {
 // TODO This currently only handles tensors with requires_grad==False correctly.
 //      It should also handle autograd.
-Operator createOperatorFromC10(const c10::OperatorHandle& op) {
+Operator createOperatorFromC10_withTracingHandledHere(const c10::OperatorHandle& op) {
```
It would be nice to have a brief comment about what the deal is with tracing handled/not handled here. Maybe just reuse text from lines 158/161/163

bhosmer commented on tools/jit/gen_jit_dispatch.py:
```diff
@@ -165,26 +172,38 @@ def from_ivalue(arg, value):
+}
+""")
+
+CONSTRUCTOR_JITONLY = CodeTemplate("""\
```
maybe one-line comments motivating the JITONLY variations.

bhosmer replied:
(specifically, the connection between 'unboxed_only' and JITONLY)

bhosmer reviewed (approved):
Some nits inline (mostly requests for comments) and tbh I think it'd be useful to take this paragraph from the PR description
> The use_c10_dispatcher setting in native_functions.yaml now has a new option 'with_codegenerated_unboxing_wrapper' which means we take the codegened unboxing wrapper from register_aten_ops.cpp and stuff it into c10. This new argument is the default, 'unboxed_only' is not the default anymore. For the (very few) ops that don't support boxed dispatch yet (i.e. ops taking TensorOptions arguments), we set them to 'unboxed_only' and they follow the old behavior of having register_aten_ops.cpp register the jit op.

and put it in `gen_jit_dispatch` where the related logic shows up.

Once we get closer to final state, I'd like us to think about names a bit - I think "boxed/unboxed" "boxing/unboxing" are a bit overworked here.

This is all minor though. If any of these suggestions would take more than a few minutes feel free to ignore, as we'll be in here again fairly soon I think.

smessmer committed:
Update on "Use codegen'ed unboxing wrappers"


Not all ops support the templated unboxing wrappers yet. For the ones that don't,
let's use the codegen'ed unboxing wrappers from register_aten_ops.cpp, but register
them with c10 directly instead of JIT.

The `use_c10_dispatcher` setting in `native_functions.yaml` now has a new option 'with_codegenerated_unboxing_wrapper' which means we take the codegened unboxing wrapper from register_aten_ops.cpp and stuff it into c10. This new argument is the default, 'unboxed_only' is not the default anymore. For the (very few) ops that don't support boxed dispatch yet (i.e. ops taking TensorOptions arguments), we set them to 'unboxed_only' and they follow the old behavior of having register_aten_ops.cpp register the jit op.

Next steps here are (1) to make TensorOptions work with boxed dispatch and remove the `unboxed_only` option from `use_c10_dispatcher`, so that all ops go through the new path and (2) make the new path template-only and remove codegen from it (see https://github.com/pytorch/pytorch/issues/32366).

Benchmarks
-------
There were some regressions visible from this originally, but after stacking this on top of https://github.com/pytorch/pytorch/pull/33732, there's no regressions anymore.

|  | add | avg1dpool | resnet |
|---|---|---|---|
| before this PR (i.e. https://github.com/pytorch/pytorch/pull/33732) | 1.66us | 5.83us | 103.61us | 
| after this PR | 1.66us | 6.09us | 103.95us | 
| after this PR and https://github.com/pytorch/pytorch/pull/33313 | | 5.74us | |

When we disable variable dispatch, we get the following numbers:

|  | add | avg1dpool | resnet |
|---|---|---|---|
| before this PR (i.e. https://github.com/pytorch/pytorch/pull/33732) | 1.55us | 4.11us | 103.36us | 
| after this PR | 1.57us | 4.23us | 102.55us | 
| after this PR and https://github.com/pytorch/pytorch/pull/33313 | | 3.84us | |


Differential Revision: [D18672405](https://our.internmc.facebook.com/intern/diff/D18672405/)

[ghstack-poisoned]

smessmer committed:
Update on "Use codegen'ed unboxing wrappers"


Not all ops support the templated unboxing wrappers yet. For the ones that don't,
let's use the codegen'ed unboxing wrappers from register_aten_ops.cpp, but register
them with c10 directly instead of JIT.

The `use_c10_dispatcher` setting in `native_functions.yaml` now has a new option 'with_codegenerated_unboxing_wrapper' which means we take the codegened unboxing wrapper from register_aten_ops.cpp and stuff it into c10. This new argument is the default, 'unboxed_only' is not the default anymore. For the (very few) ops that don't support boxed dispatch yet (i.e. ops taking TensorOptions arguments), we set them to 'unboxed_only' and they follow the old behavior of having register_aten_ops.cpp register the jit op.

Next steps here are (1) to make TensorOptions work with boxed dispatch and remove the `unboxed_only` option from `use_c10_dispatcher`, so that all ops go through the new path and (2) make the new path template-only and remove codegen from it (see https://github.com/pytorch/pytorch/issues/32366).

Benchmarks
-------
There were some regressions visible from this originally, but after stacking this on top of https://github.com/pytorch/pytorch/pull/33732, there's no regressions anymore.

|  | add | avg1dpool | resnet |
|---|---|---|---|
| before this PR (i.e. https://github.com/pytorch/pytorch/pull/33732) | 1.66us | 5.83us | 103.61us | 
| after this PR | 1.66us | 6.09us | 103.95us | 
| after this PR and https://github.com/pytorch/pytorch/pull/33313 | | 5.74us | |

When we disable variable dispatch, we get the following numbers:

|  | add | avg1dpool | resnet |
|---|---|---|---|
| before this PR (i.e. https://github.com/pytorch/pytorch/pull/33732) | 1.55us | 4.11us | 103.36us | 
| after this PR | 1.57us | 4.23us | 102.55us | 
| after this PR and https://github.com/pytorch/pytorch/pull/33313 | | 3.84us | |


Differential Revision: [D18672405](https://our.internmc.facebook.com/intern/diff/D18672405/)

[ghstack-poisoned]

smessmer committed:
Update on "Use codegen'ed unboxing wrappers"


Not all ops support the templated unboxing wrappers yet. For the ones that don't,
let's use the codegen'ed unboxing wrappers from register_aten_ops.cpp, but register
them with c10 directly instead of JIT.

The `use_c10_dispatcher` setting in `native_functions.yaml` now has a new option 'with_codegenerated_unboxing_wrapper' which means we take the codegened unboxing wrapper from register_aten_ops.cpp and stuff it into c10. This new argument is the default, 'unboxed_only' is not the default anymore. For the (very few) ops that don't support boxed dispatch yet (i.e. ops taking TensorOptions arguments), we set them to 'unboxed_only' and they follow the old behavior of having register_aten_ops.cpp register the jit op.

Next steps here are (1) to make TensorOptions work with boxed dispatch and remove the `unboxed_only` option from `use_c10_dispatcher`, so that all ops go through the new path and (2) make the new path template-only and remove codegen from it (see https://github.com/pytorch/pytorch/issues/32366).

Benchmarks
-------
There were some regressions visible from this originally, but after stacking this on top of https://github.com/pytorch/pytorch/pull/33732, there's no regressions anymore.

|  | add | avg1dpool | resnet |
|---|---|---|---|
| before this PR (i.e. https://github.com/pytorch/pytorch/pull/33732) | 1.66us | 5.83us | 103.61us | 
| after this PR | 1.66us | 6.09us | 103.95us | 
| after this PR and https://github.com/pytorch/pytorch/pull/33313 | | 5.74us | |

When we disable variable dispatch, we get the following numbers:

|  | add | avg1dpool | resnet |
|---|---|---|---|
| before this PR (i.e. https://github.com/pytorch/pytorch/pull/33732) | 1.55us | 4.11us | 103.36us | 
| after this PR | 1.57us | 4.23us | 102.55us | 
| after this PR and https://github.com/pytorch/pytorch/pull/33313 | | 3.84us | |


Differential Revision: [D18672405](https://our.internmc.facebook.com/intern/diff/D18672405/)

[ghstack-poisoned]

smessmer committed:
Update on "Use codegen'ed unboxing wrappers"


Not all ops support the templated unboxing wrappers yet. For the ones that don't,
let's use the codegen'ed unboxing wrappers from register_aten_ops.cpp, but register
them with c10 directly instead of JIT.

The `use_c10_dispatcher` setting in `native_functions.yaml` now has a new option 'with_codegenerated_unboxing_wrapper' which means we take the codegened unboxing wrapper from register_aten_ops.cpp and stuff it into c10. This new argument is the default, 'unboxed_only' is not the default anymore. For the (very few) ops that don't support boxed dispatch yet (i.e. ops taking TensorOptions arguments), we set them to 'unboxed_only' and they follow the old behavior of having register_aten_ops.cpp register the jit op.

Next steps here are (1) to make TensorOptions work with boxed dispatch and remove the `unboxed_only` option from `use_c10_dispatcher`, so that all ops go through the new path and (2) make the new path template-only and remove codegen from it (see https://github.com/pytorch/pytorch/issues/32366).

Benchmarks
-------
There were some regressions visible from this originally, but after stacking this on top of https://github.com/pytorch/pytorch/pull/33732, there's no regressions anymore.

|  | add | avg1dpool | resnet |
|---|---|---|---|
| before this PR (i.e. https://github.com/pytorch/pytorch/pull/33732) | 1.66us | 5.83us | 103.61us | 
| after this PR | 1.66us | 6.09us | 103.95us | 
| after this PR and https://github.com/pytorch/pytorch/pull/33313 | | 5.74us | |

When we disable variable dispatch, we get the following numbers:

|  | add | avg1dpool | resnet |
|---|---|---|---|
| before this PR (i.e. https://github.com/pytorch/pytorch/pull/33732) | 1.55us | 4.11us | 103.36us | 
| after this PR | 1.57us | 4.23us | 102.55us | 
| after this PR and https://github.com/pytorch/pytorch/pull/33313 | | 3.84us | |


Differential Revision: [D18672405](https://our.internmc.facebook.com/intern/diff/D18672405/)

[ghstack-poisoned]

smessmer committed:
Update on "Use codegen'ed unboxing wrappers"


Not all ops support the templated unboxing wrappers yet. For the ones that don't,
let's use the codegen'ed unboxing wrappers from register_aten_ops.cpp, but register
them with c10 directly instead of JIT.

The `use_c10_dispatcher` setting in `native_functions.yaml` now has a new option 'with_codegenerated_unboxing_wrapper' which means we take the codegened unboxing wrapper from register_aten_ops.cpp and stuff it into c10. This new argument is the default, 'unboxed_only' is not the default anymore. For the (very few) ops that don't support boxed dispatch yet (i.e. ops taking TensorOptions arguments), we set them to 'unboxed_only' and they follow the old behavior of having register_aten_ops.cpp register the jit op.

Next steps here are (1) to make TensorOptions work with boxed dispatch and remove the `unboxed_only` option from `use_c10_dispatcher`, so that all ops go through the new path and (2) make the new path template-only and remove codegen from it (see https://github.com/pytorch/pytorch/issues/32366).

Benchmarks
-------
There were some regressions visible from this originally, but after stacking this on top of https://github.com/pytorch/pytorch/pull/33732, there's no regressions anymore.

|  | add | avg1dpool | resnet |
|---|---|---|---|
| before this PR (i.e. https://github.com/pytorch/pytorch/pull/33732) | 1.66us | 5.83us | 103.61us | 
| after this PR | 1.66us | 6.09us | 103.95us | 
| after this PR and https://github.com/pytorch/pytorch/pull/33313 | | 5.74us | |

When we disable variable dispatch, we get the following numbers:

|  | add | avg1dpool | resnet |
|---|---|---|---|
| before this PR (i.e. https://github.com/pytorch/pytorch/pull/33732) | 1.55us | 4.11us | 103.36us | 
| after this PR | 1.57us | 4.23us | 102.55us | 
| after this PR and https://github.com/pytorch/pytorch/pull/33313 | | 3.84us | |


Differential Revision: [D18672405](https://our.internmc.facebook.com/intern/diff/D18672405/)

[ghstack-poisoned]

smessmer committed:
Update on "Use codegen'ed unboxing wrappers"


Not all ops support the templated unboxing wrappers yet. For the ones that don't,
let's use the codegen'ed unboxing wrappers from register_aten_ops.cpp, but register
them with c10 directly instead of JIT.

The `use_c10_dispatcher` setting in `native_functions.yaml` now has a new option 'with_codegenerated_unboxing_wrapper' which means we take the codegened unboxing wrapper from register_aten_ops.cpp and stuff it into c10. This new argument is the default, 'unboxed_only' is not the default anymore. For the (very few) ops that don't support boxed dispatch yet (i.e. ops taking TensorOptions arguments), we set them to 'unboxed_only' and they follow the old behavior of having register_aten_ops.cpp register the jit op.

Next steps here are (1) to make TensorOptions work with boxed dispatch and remove the `unboxed_only` option from `use_c10_dispatcher`, so that all ops go through the new path and (2) make the new path template-only and remove codegen from it (see https://github.com/pytorch/pytorch/issues/32366).

Benchmarks
-------
There were some regressions visible from this originally, but after stacking this on top of https://github.com/pytorch/pytorch/pull/33732, there's no regressions anymore.

|  | add | avg1dpool | resnet |
|---|---|---|---|
| before this PR (i.e. https://github.com/pytorch/pytorch/pull/33732) | 1.66us | 5.83us | 103.61us | 
| after this PR | 1.66us | 6.09us | 103.95us | 
| after this PR and https://github.com/pytorch/pytorch/pull/33313 | | 5.74us | |

When we disable variable dispatch, we get the following numbers:

|  | add | avg1dpool | resnet |
|---|---|---|---|
| before this PR (i.e. https://github.com/pytorch/pytorch/pull/33732) | 1.55us | 4.11us | 103.36us | 
| after this PR | 1.57us | 4.23us | 102.55us | 
| after this PR and https://github.com/pytorch/pytorch/pull/33313 | | 3.84us | |


Differential Revision: [D18672405](https://our.internmc.facebook.com/intern/diff/D18672405/)

[ghstack-poisoned]

smessmer committed:
Update on "Use codegen'ed unboxing wrappers"


Not all ops support the templated unboxing wrappers yet. For the ones that don't,
let's use the codegen'ed unboxing wrappers from register_aten_ops.cpp, but register
them with c10 directly instead of JIT.

The `use_c10_dispatcher` setting in `native_functions.yaml` now has a new option 'with_codegenerated_unboxing_wrapper' which means we take the codegened unboxing wrapper from register_aten_ops.cpp and stuff it into c10. This new argument is the default, 'unboxed_only' is not the default anymore. For the (very few) ops that don't support boxed dispatch yet (i.e. ops taking TensorOptions arguments), we set them to 'unboxed_only' and they follow the old behavior of having register_aten_ops.cpp register the jit op.

Next steps here are (1) to make TensorOptions work with boxed dispatch and remove the `unboxed_only` option from `use_c10_dispatcher`, so that all ops go through the new path and (2) make the new path template-only and remove codegen from it (see https://github.com/pytorch/pytorch/issues/32366).

Benchmarks
-------
There were some regressions visible from this originally, but after stacking this on top of https://github.com/pytorch/pytorch/pull/33732, there's no regressions anymore.

|  | add | avg1dpool | resnet |
|---|---|---|---|
| before this PR (i.e. https://github.com/pytorch/pytorch/pull/33732) | 1.66us | 5.83us | 103.61us | 
| after this PR | 1.66us | 6.09us | 103.95us | 
| after this PR and https://github.com/pytorch/pytorch/pull/33313 | | 5.74us | |

When we disable variable dispatch, we get the following numbers:

|  | add | avg1dpool | resnet |
|---|---|---|---|
| before this PR (i.e. https://github.com/pytorch/pytorch/pull/33732) | 1.55us | 4.11us | 103.36us | 
| after this PR | 1.57us | 4.23us | 102.55us | 
| after this PR and https://github.com/pytorch/pytorch/pull/33313 | | 3.84us | |


Differential Revision: [D18672405](https://our.internmc.facebook.com/intern/diff/D18672405/)

[ghstack-poisoned]

smessmer committed:
Update on "Use codegen'ed unboxing wrappers"


Not all ops support the templated unboxing wrappers yet. For the ones that don't,
let's use the codegen'ed unboxing wrappers from register_aten_ops.cpp, but register
them with c10 directly instead of JIT.

The `use_c10_dispatcher` setting in `native_functions.yaml` now has a new option 'with_codegenerated_unboxing_wrapper' which means we take the codegened unboxing wrapper from register_aten_ops.cpp and stuff it into c10. This new argument is the default, 'unboxed_only' is not the default anymore. For the (very few) ops that don't support boxed dispatch yet (i.e. ops taking TensorOptions arguments), we set them to 'unboxed_only' and they follow the old behavior of having register_aten_ops.cpp register the jit op.

Next steps here are (1) to make TensorOptions work with boxed dispatch and remove the `unboxed_only` option from `use_c10_dispatcher`, so that all ops go through the new path and (2) make the new path template-only and remove codegen from it (see https://github.com/pytorch/pytorch/issues/32366).

Benchmarks
-------
There were some regressions visible from this originally, but after stacking this on top of https://github.com/pytorch/pytorch/pull/33732, there's no regressions anymore.

|  | add | avg1dpool | resnet |
|---|---|---|---|
| before this PR (i.e. https://github.com/pytorch/pytorch/pull/33732) | 1.66us | 5.83us | 103.61us | 
| after this PR | 1.66us | 6.09us | 103.95us | 
| after this PR and https://github.com/pytorch/pytorch/pull/33313 | | 5.74us | |

When we disable variable dispatch, we get the following numbers:

|  | add | avg1dpool | resnet |
|---|---|---|---|
| before this PR (i.e. https://github.com/pytorch/pytorch/pull/33732) | 1.55us | 4.11us | 103.36us | 
| after this PR | 1.57us | 4.23us | 102.55us | 
| after this PR and https://github.com/pytorch/pytorch/pull/33313 | | 3.84us | |


Differential Revision: [D18672405](https://our.internmc.facebook.com/intern/diff/D18672405/)

[ghstack-poisoned]

smessmer committed:
Update on "Use codegen'ed unboxing wrappers"


Not all ops support the templated unboxing wrappers yet. For the ones that don't,
let's use the codegen'ed unboxing wrappers from register_aten_ops.cpp, but register
them with c10 directly instead of JIT.

The `use_c10_dispatcher` setting in `native_functions.yaml` now has a new option 'with_codegenerated_unboxing_wrapper' which means we take the codegened unboxing wrapper from register_aten_ops.cpp and stuff it into c10. This new argument is the default, 'unboxed_only' is not the default anymore. For the (very few) ops that don't support boxed dispatch yet (i.e. ops taking TensorOptions arguments), we set them to 'unboxed_only' and they follow the old behavior of having register_aten_ops.cpp register the jit op.

Next steps here are (1) to make TensorOptions work with boxed dispatch and remove the `unboxed_only` option from `use_c10_dispatcher`, so that all ops go through the new path and (2) make the new path template-only and remove codegen from it (see https://github.com/pytorch/pytorch/issues/32366).

Benchmarks
-------
There were some regressions visible from this originally, but after stacking this on top of https://github.com/pytorch/pytorch/pull/33732, there's no regressions anymore.

|  | add | avg1dpool | resnet |
|---|---|---|---|
| before this PR (i.e. https://github.com/pytorch/pytorch/pull/33732) | 1.66us | 5.83us | 103.61us | 
| after this PR | 1.66us | 6.09us | 103.95us | 
| after this PR and https://github.com/pytorch/pytorch/pull/33313 | | 5.74us | |

When we disable variable dispatch, we get the following numbers:

|  | add | avg1dpool | resnet |
|---|---|---|---|
| before this PR (i.e. https://github.com/pytorch/pytorch/pull/33732) | 1.55us | 4.11us | 103.36us | 
| after this PR | 1.57us | 4.23us | 102.55us | 
| after this PR and https://github.com/pytorch/pytorch/pull/33313 | | 3.84us | |


Differential Revision: [D18672405](https://our.internmc.facebook.com/intern/diff/D18672405/)

[ghstack-poisoned]

ljk53 commented:
Reverting this PR as it broke mobile CI (static dispatch) and I failed to find any quick fix: https://github.com/pytorch/pytorch/pull/34013#issuecomment-593003791

ljk53 commented on aten/src/ATen/core/dispatch/DispatchTable.h:
```diff
@@ -188,12 +198,32 @@ class DispatchTable final {
+    TORCH_INTERNAL_ASSERT(!manuallyBoxedKernel_.has_value(), "Cannot set multiple manually boxed kernels for the same operator ", operatorName_);
+    manuallyBoxedKernel_ = func;
+
+    // make sure that all previously registered kernels get this manually boxed kernel
```
Should we also check previously registered catchAllKernel_?
https://github.com/ljk53/pytorch/commit/2d7ede5f71693dca308b00570dcaee6c4148864e - this fixed a crash I ran into with mobile build.

bhosmer replied:
Sounds like this needs a rebase - catch-all kernels went away in #33817 (reducing the possibility this sort of bug being one of the motivations for the simplification)

smessmer replied:
They didn't go away entirely, DispatchTable.h still has them.

@ljk53 : good catch, I added it to https://github.com/pytorch/pytorch/pull/34177
